### ISSUE TITLE
xSQLServerSetup: xSQLServerSetup and ConfigurationFile support

### ISSUE THREAD

**Issue by nilsvbsol** [PowerShell/xSQLServer#377]
# Improvement suggested:

- In the **xSQLServerSetup**-resource: it's nice to have a lot of parameters at hand for installing SQL Server, but one is missing in my opinion, namely "ConfigurationFile".
See the following article: http://www.colinsalmcorner.com/post/install-and-configure-sql-server-using-powershell-dsc
In this article the configuration-file of the SQL install-wizard is used to install SQL-server.
It would be nice if this resource could also work with this configuration-file. It is in fact suggested at MSDN as well: https://msdn.microsoft.com/en-us/library/dd239405.aspx

**Comment by randomnote1** [comment]
In my opinion the configuration file is irrelevant when using DSC. DSC houses all of the data the configuration file does, plus it's dynamic.

**Comment by johlju** [comment]
@nilsvbsol using the configuration file I don't see that it would be possible to check that the configuration is in desired state. Do you maybe thought of it and have a suggestion to accomplish that? Do you have a production scenario that this would be beneficial in?

**Comment by nilsvbsol** [comment]
@randomnote1: yes, that's true. But the configfile has been a way to install SQL-Server unattended for years. DSC should be consisted with existing methods that leverage unattended installs in my opinion. Especially if this method is still openly supported by Microsoft.
@johlju: In my opinion, the configurationfile is a fast and convenient way to house the config of SQLServer. As DSC-scripts can comprise of several 'domains' (=e.g. components for pure software development, components for server infrastructure, user experience, etc...) they can quickly become too big and disordered.

As I see it, leaving the config in the configfile addresses several **advantages**:
- The configfile is an OEM-asset (e.g. it's composed by Microsoft and **is always 100% valid**). Converting from one format to the other (file -> DSC) may be complex for someone who doesn't know SQL in-depth, aka the chance for errors or making conflicting settings is big (some parameters exclude each other), whereas the file is valid and usable _out of the box_ (*).
- DBA's may have several config-files lying around, collected over the years. These files would be used to deploy several types of servers and those configs may have proven their worth already. DBA's would be confident to re-use them on new servers. Porting them over to the _disconnected_ DSC-language may prove to be a step too far for DBA's.
If xSQLServerSetup could make use of those config-files, a DBA could easily hand over the configfile to a DevOps guy, who can install SQL-server via DSC in minutes.

(*): _almost_ out-of-the-box, since the configfile generated via the SQL install-wizard sets the UIMode on standard and Quiet on false, aka: during DSC-setup it will fail as dialogboxes will be shown, but that's easily corrected.

The added value of **xSQLServerSetup** in this story could be that it implicitly overrides certain parameters stated in the configfile (e.g. UIMode and Quiet/QuitSimple parameters).

How would xSQLServerSetup work in this way?
It should map config-file parameters to the existing DSC-parameters (most of them are one-on-one mappable).

As for checking for a 'desired state' goes: if it could read all the parameters of the configfile and knows how to map them to the DSC ones, it could make use of the current checks to ensure a desired state.
**It would just be an extension to the current flow, extended with 1 extra mapping step.**

**Comment by randomnote1** [comment]
@Nilsvbsol, that's a very well thought out response. Thanks!

The tricky part about this whole thing is that the INI would need to be copied to and maintained on the SQL server (or to a location the PSDSCCredential has access to) for the lifespan of the instance.

Perhaps a helper function could be written and included with the module to convert an INI to a DSC configuration. This has its own challenges from the perspective of handling standalone vs cluster installations.

I guess the real question is, if the project decides to support configuration files, where will it be supported? During the get/set/test operation? Or during the initial compilation?

**Comment by Zuldan** [comment]
@nilsvbsol, xSQLServerSetup is another form of an unattended install. It does make sense to have an unattended install system use another unattended install system to install SQL.

**Comment by johlju** [comment]
Most what I say here has already been mentioned, but I just scribbled down my thoughts.

My thought was, as @randomnote1 mentioned, to have a method of converting the .ini file to the correct DSC configuration. But not as part of xSQLServerSetup, but as a standalone method. 

But if there is more interest of having the .ini built inside the xSQLServerSetup resource in a way similar to what @nilsvbsol suggest then we have to figure out, as @randomnote1 points out, how it is practically gonna work.
I guess one parameter 'ConfigurationFile' could point to a UNC path (or optionally be copied locally), and override all other parameters in the DSC configuration.
I also guess both the Get-method and the Set-method could both call one or more helper function to  get the "mapping hash table" depending on the installation scenario. Test-method will probably utilize the logic in the Get-method to evaluate if the system is in desired state.
But all this, I'm afraid this is gonna be _a lot_ of code logic. 😄
